# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

## Instructions:

The steps for this challenge are broken out into the following sections:

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model 

Using your knowledge of Pandas and scikit-learn’s `StandardScaler()`, preprocess the dataset so that you can use it to compile and evaluate the neural network model later.

Open the starter code file, and complete the following data preparation steps:

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.   

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.
 
3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use your knowledge of TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset. Consider the number of inputs before determining the number of layers that your model will contain or the number of neurons on each layer. Then, compile and fit your model. Finally, evaluate your binary classification model to calculate the model’s loss and accuracy. 
 
To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

> **Hint** When fitting the model, start with a small number of epochs, such as 20, 50, or 100.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

4. Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 

### Optimize the Neural Network Model

Using your knowledge of TensorFlow and Keras, optimize your model to improve the model's accuracy. Even if you do not successfully achieve a better accuracy, you'll need to demonstrate at least two attempts to optimize the model. You can include these attempts in your existing notebook. Or, you can make copies of the starter notebook in the same folder, rename them, and code each model optimization in a new notebook. 

> **Note** You will not lose points if your model does not achieve a high accuracy, as long as you make at least two attempts to optimize the model.

To do so, complete the following steps:

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. After finishing your models, display the accuracy scores achieved by each model, and compare the results.

3. Save each of your models as an HDF5 file.


In [1]:
# Imports
import pandas as pd
from pathlib import Path
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder

2022-12-14 18:44:08.955118: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-12-14 18:44:09.079477: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-12-14 18:44:09.082236: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-12-14 18:44:09.082248: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore 

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.  


In [2]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
applicant_data_df = pd.read_csv(Path('./Resources/applicants_data.csv'))

# Review the DataFrame
applicant_data_df.head()


EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Review the data types associated with the columns
applicant_data_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   EIN                     34299 non-null  int64 
 1   NAME                    34299 non-null  object
 2   APPLICATION_TYPE        34299 non-null  object
 3   AFFILIATION             34299 non-null  object
 4   CLASSIFICATION          34299 non-null  object
 5   USE_CASE                34299 non-null  object
 6   ORGANIZATION            34299 non-null  object
 7   STATUS                  34299 non-null  int64 
 8   INCOME_AMT              34299 non-null  object
 9   SPECIAL_CONSIDERATIONS  34299 non-null  object
 10  ASK_AMT                 34299 non-null  int64 
 11  IS_SUCCESSFUL           34299 non-null  int64 
dtypes: int64(4), object(8)
memory usage: 3.1+ MB


### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [4]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame
applicant_data_df = applicant_data_df.drop(columns=['EIN', 'NAME'])

# Review the DataFrame
applicant_data_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   APPLICATION_TYPE        34299 non-null  object
 1   AFFILIATION             34299 non-null  object
 2   CLASSIFICATION          34299 non-null  object
 3   USE_CASE                34299 non-null  object
 4   ORGANIZATION            34299 non-null  object
 5   STATUS                  34299 non-null  int64 
 6   INCOME_AMT              34299 non-null  object
 7   SPECIAL_CONSIDERATIONS  34299 non-null  object
 8   ASK_AMT                 34299 non-null  int64 
 9   IS_SUCCESSFUL           34299 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 2.6+ MB


### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [5]:
# Create a list of categorical variables 
categorical_variables = list(applicant_data_df.select_dtypes(include=['object']).columns)

# Display the categorical variables list
categorical_variables


['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT',
 'SPECIAL_CONSIDERATIONS']

In [6]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)


In [7]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])


In [8]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(
    encoded_data,
    columns = enc.get_feature_names_out(categorical_variables)
)

# Review the DataFrame
encoded_df.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                   0.0  ...                0.0                     0.0   
1                   0.0  ...                1.0                     0.0   
2                   0.0  ...                0.0                     0.0   
3                   0.0  ...                0.0                     1.0   
4                   0.0  ...                0.0                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                       1.0                       0.0  
1                       1.0                       0.0  
2                       1.0                       0.0  
3                       1.0                       0.0  
4                       1.0                       0.0  

[5 rows x 114 columns]

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

In [9]:
# Add the numerical variables from the original DataFrame to the one-hot encoding DataFrame
encoded_df = pd.concat(
    [
        encoded_df, 
        applicant_data_df[
            list(applicant_data_df.select_dtypes(exclude=['object']).columns)
        ]
    ], 
    axis=1
)

# Review the Dataframe
display(encoded_df.head())
display(encoded_df.tail())

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                   0.0  ...                 0.0               0.0   
1                   0.0  ...                 0.0               0.0   
2                   0.0  ...                 0.0               0.0   
3                   0.0  ...                 0.0               0.0   
4                   0.0  ...                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  STATUS  ASK_AMT  \
0                       1.0                       0.0       1     5000   
1                       1.0                       0.0       1   108590   
2                       1.0                       0.0       1     5000   
3                       1.0                       0.0       1     6692   
4                       1.0                       0.0       1   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1  

[5 rows x 117 columns]

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
34294                   0.0                  0.0                   0.0   
34295                   0.0                  0.0                   0.0   
34296                   0.0                  0.0                   0.0   
34297                   0.0                  0.0                   0.0   
34298                   0.0                  0.0                   0.0   

       APPLICATION_TYPE_T29  ...  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
34294                   0.0  ...                 0.0               0.0   
34295                   0.0  ...                 0.0               0.0   
34296                   0.0  ...                 0.0               0.0   
34297                   0.0  ...                 0.0               0.0   
34298                   0.0  ...                 0.0               1.0   

       INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
34294                     0.0              0.0                0.0   
34295                     0.0              0.0                0.0   
34296                     0.0              0.0                0.0   
34297                     0.0              0.0                0.0   
34298                     0.0              0.0                0.0   

       SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  STATUS   ASK_AMT  \
34294                       1.0                       0.0       1      5000   
34295                       1.0                       0.0       1      5000   
34296                       1.0                       0.0       1      5000   
34297                       1.0                       0.0       1      5000   
34298                       1.0                       0.0       1  36500179   

       IS_SUCCESSFUL  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[5 rows x 117 columns]

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 



In [10]:
# Define the target set y using the IS_SUCCESSFUL column
y = encoded_df['IS_SUCCESSFUL']

# Display a sample of y
y[:5]


0    1
1    1
2    0
3    1
4    1
Name: IS_SUCCESSFUL, dtype: int64

In [11]:
# Define features set X by selecting all columns but IS_SUCCESSFUL
X = encoded_df.drop(columns=['IS_SUCCESSFUL'])

# Review the features DataFrame
X.head()


APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                   0.0  ...                       0.0                 0.0   
1                   0.0  ...                       0.0                 0.0   
2                   0.0  ...                       0.0                 0.0   
3                   0.0  ...                       0.0                 0.0   
4                   0.0  ...                       1.0                 0.0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0               0.0                     0.0              0.0   
1               0.0                     0.0              0.0   
2               0.0                     0.0              0.0   
3               0.0                     0.0              0.0   
4               0.0                     0.0              0.0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  \
0                0.0                       1.0                       0.0   
1                0.0                       1.0                       0.0   
2                0.0                       1.0                       0.0   
3                0.0                       1.0                       0.0   
4                0.0                       1.0                       0.0   

   STATUS  ASK_AMT  
0       1     5000  
1       1   108590  
2       1     5000  
3       1     6692  
4       1   142590  

[5 rows x 116 columns]

### Step 6: Split the features and target sets into training and testing datasets.


In [12]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [13]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [14]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.columns)

# Review the number of features
number_input_features


116

In [15]:
# Define the number of neurons in the output layer
number_output_neurons = 1

In [16]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 =  (number_input_features + 1) // number_output_neurons

# Review the number hidden nodes in the first layer
hidden_nodes_layer1


117

In [17]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 =  (hidden_nodes_layer1 + 1) // 2 

# Review the number hidden nodes in the second layer
hidden_nodes_layer2


59

In [18]:
# Create the Sequential model instance
nn = Sequential()


2022-12-14 18:44:10.533645: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-12-14 18:44:10.533690: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2022-12-14 18:44:10.533704: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (Fireball): /proc/driver/nvidia/version does not exist
2022-12-14 18:44:10.533870: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [19]:
# Add the first hidden layer
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))


In [20]:
# Add the second hidden layer
nn.add(Dense(units=hidden_nodes_layer2, activation="relu"))


In [21]:
# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(Dense(units=number_output_neurons, activation="sigmoid"))


In [22]:
# Display the Sequential model summary
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 117)               13689     
                                                                 
 dense_1 (Dense)             (None, 59)                6962      
                                                                 
 dense_2 (Dense)             (None, 1)                 60        
                                                                 
Total params: 20,711
Trainable params: 20,711
Non-trainable params: 0
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [23]:
# Compile the Sequential model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [24]:
# Fit the model using 50 epochs and the training data
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)


Epoch 1/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5721 - accuracy: 0.7232
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5550 - accuracy: 0.7294
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5496 - accuracy: 0.7318
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5472 - accuracy: 0.7328
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5467 - accuracy: 0.7324
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5445 - accuracy: 0.7334
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5441 - accuracy: 0.7350
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5428 - accuracy: 0.7346
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5422 - accuracy: 0.7349
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5421 - accura

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [25]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5729 - accuracy: 0.7305 - 295ms/epoch - 1ms/step
Loss: 0.5728937387466431, Accuracy: 0.7304956316947937


### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 


In [26]:
# Set the model's file path
file_path = Path("Resources/AlphabetSoup.h5")

# Export your model to a HDF5 file
nn.save(file_path)


---

## Optimize the neural network model


### Step 1: Define at least three new deep neural network models (resulting in the original plus 3 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.


### Alternative Model 1

In [27]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.columns)

# Review the number of features
number_input_features

116

In [28]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

In [29]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 = hidden_nodes_layer1 * 1.5

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A1

175.5

In [30]:
# Adding 2 more hidden layers
hidden_nodes_layer2_A1 =  (hidden_nodes_layer1_A1 + 1) // 2 
hidden_nodes_layer3_A1 =  (hidden_nodes_layer2_A1 + 1) // 2 

# Review the number hidden nodes in the second layer
display(hidden_nodes_layer2_A1)
display(hidden_nodes_layer3_A1)

88.0

44.0

In [31]:
# Create the Sequential model instance
nn_A1 = Sequential()

In [32]:
# First, second and third hidden layers
nn_A1.add(Dense(units=hidden_nodes_layer1_A1, input_dim=number_input_features, activation="relu"))
nn_A1.add(Dense(units=hidden_nodes_layer2_A1, activation="relu"))
nn_A1.add(Dense(units=hidden_nodes_layer3_A1, activation="relu"))

# Output layer
nn_A1.add(Dense(units=number_output_neurons_A1, activation="sigmoid"))


# Check the structure of the model
nn_A1.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 175)               20475     
                                                                 
 dense_4 (Dense)             (None, 88)                15488     
                                                                 
 dense_5 (Dense)             (None, 44)                3916      
                                                                 
 dense_6 (Dense)             (None, 1)                 45        
                                                                 
Total params: 39,924
Trainable params: 39,924
Non-trainable params: 0
_________________________________________________________________


In [33]:
# Compile the Sequential model
nn_A1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [34]:
# Fit the model using 50 epochs and the training data
fit_model_A1 = nn_A1.fit(X_train_scaled, y_train, epochs=100)


Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.5688 - accuracy: 0.7240
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5525 - accuracy: 0.7297
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5496 - accuracy: 0.7314
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5463 - accuracy: 0.7320
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5464 - accuracy: 0.7343
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5444 - accuracy: 0.7343
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5434 - accuracy: 0.7349
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5429 - accuracy: 0.7361
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5417 - accuracy: 0.7361
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5416 - accura

In [35]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss_A1, model_accuracy_A1 = nn_A1.evaluate(X_test_scaled, y_test,verbose=2)

# Display the model loss and accuracy results
print(f"A1 Loss: {model_loss_A1}, A1 Accuracy: {model_accuracy_A1}")

268/268 - 0s - loss: 0.5899 - accuracy: 0.7280 - 374ms/epoch - 1ms/step
A1 Loss: 0.5899295210838318, A1 Accuracy: 0.7280466556549072


#### Alternative Model 2

In [36]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

116

In [37]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [38]:
# For alternative 3, we will use 2 hidden layers with more neurons
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A2 = number_input_features * 2

# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A2 = hidden_nodes_layer1_A2

# Review the number of hidden nodes in the first layer
display(hidden_nodes_layer1_A2)
display(hidden_nodes_layer2_A2)

232

232

In [39]:
# Create the Sequential model instance
nn_A2 = Sequential()

In [40]:
# First hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer1_A2, input_dim=number_input_features, activation="relu"))
nn_A2.add(Dense(units=hidden_nodes_layer2_A2, activation="relu"))

# Output layer
nn_A2.add(Dense(units=number_output_neurons_A2, activation="tanh"))

# Check the structure of the model
nn_A2.summary()


Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 232)               27144     
                                                                 
 dense_8 (Dense)             (None, 232)               54056     
                                                                 
 dense_9 (Dense)             (None, 1)                 233       
                                                                 
Total params: 81,433
Trainable params: 81,433
Non-trainable params: 0
_________________________________________________________________


In [41]:
# Compile the model
nn_A2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [42]:
# Fit the model
fit_model_A2 = nn_A2.fit(X_train_scaled, y_train, epochs=100)


Epoch 1/100
804/804 [==============================] - 1s 1ms/step - loss: 0.8035 - accuracy: 0.7085
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6282 - accuracy: 0.7216
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6096 - accuracy: 0.7246
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6111 - accuracy: 0.7243
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5838 - accuracy: 0.7297
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5887 - accuracy: 0.7288
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5840 - accuracy: 0.7310
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5762 - accuracy: 0.7305
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5924 - accuracy: 0.7314
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6005 - accura

### Step 2: After finishing your models, display the accuracy scores achieved by each model, and compare the results.

In [43]:
print("Original Model Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Original Model Results
268/268 - 0s - loss: 0.5729 - accuracy: 0.7305 - 199ms/epoch - 743us/step
Loss: 0.5728937387466431, Accuracy: 0.7304956316947937


In [44]:
print("Alternative Model 1 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss_A1, model_accuracy_A1 = nn_A1.evaluate(X_test_scaled, y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss A1: {model_loss_A1}, Accuracy A1: {model_accuracy_A1}")

Alternative Model 1 Results
268/268 - 0s - loss: 0.5899 - accuracy: 0.7280 - 209ms/epoch - 778us/step
Loss A1: 0.5899295210838318, Accuracy A1: 0.7280466556549072


In [45]:
print("Alternative Model 2 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss_A2, model_accuracy_A2 = nn_A2.evaluate(X_test_scaled, y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss A2: {model_loss_A2}, Accuracy A2: {model_accuracy_A2}")

Alternative Model 2 Results
268/268 - 0s - loss: 0.6524 - accuracy: 0.7278 - 293ms/epoch - 1ms/step
Loss A2: 0.6523727178573608, Accuracy A2: 0.7278134226799011


### Step 3: Save each of your alternative models as an HDF5 file.


In [46]:
# Set the file path for the first alternative model
file_path = Path("Resources/AlphabetSoup_A1.h5")

# Export your model to a HDF5 file
nn_A1.save(file_path)


In [47]:
# Set the file path for the second alternative model
file_path = Path("Resources/AlphabetSoup_A2.h5")

# Export your model to a HDF5 file
nn_A2.save(file_path)


### Observations
Our `original` model, using `1` hidden layer with `59` nodes, `Rectified Linear Unit` activation function and its output layer using the `sigmoid` activation function achieved an accuracy at `0.73049` and loss of `0.57289` performed better than either `Alternative 1` with `2` hidden layers, each with `88` and `44` nodes and the `sigmoid` activation fuction in its output layer resulting in an accuracy of `0.72804` and a loss of `0.58992` or even `Alternative 3` with `1` hidden layer with `232` nodes and `tanh` as the output layer activation function yielding an accuracy of `0.72781` and a loss of `0.65237`
